### PR TITLE
Calypso Build Package: Make `WP` flag add JSX -> `@wordpress/element` transforms

### PR DESCRIPTION
**Pull request by ockham**
Per Jon's suggestion.

#### Changes proposed in this Pull Request

* Turn calypso-build's wordpress-element Babel preset into a function
  * I was getting `Plugin/Preset files are not allowed to export objects, only functions` :man_shrugging: 
* Change `calypso-build/babel.config.js` into an actual Babel preset, `calypso-build/babel/default.js`. This allows our webpack config to pass a list of presets (whose elements depend on the `WP` flag) to the Babel loader, without needing to worry if they're going to collide with a Babel config.
* Calypso Build Package: Make `WP` flag add JSX -> `@wordpress/element` transforms

#### Testing instructions

Do the following both on `master`, and this branch:

```
npm run distclean && npm ci
npx lerna run build
```

Keep the resulting `apps/*/dist/` directories, and `diff` them. They should be identical.

Furthermore, verify that Calypso still builds and runs (`npm start`).
Finally, verify that it's still possible to build the client in production mode, for the `fallback` target: `npm run build-client-fallback`.

### CODE REVIEW

**Comment by ockham** [comment]
We might need to change this line: https://github.com/Automattic/wp-calypso/pull/32601/files#diff-b99af7f62e015cd796a25e0d65088f6aR63

...from `process.cwd()` to some sort of Webpack context. I think this is what's currently causing the `build-notifications` task to fail, since it's run from the Calypso root dir, where there _is_ a `babel.config.js` one (but we don't want to use this one).

**Comment by ockham** [comment]
> We might need to change this line: https://github.com/Automattic/wp-calypso/pull/32601/files#diff-b99af7f62e015cd796a25e0d65088f6aR63
> 
> ...from `process.cwd()` to some sort of Webpack context. I think this is what's currently causing the `build-notifications` task to fail, since it's run from the Calypso root dir, where there _is_ a `babel.config.js` one (but we don't want to use this one).

Decided to not add a workaround for that since it's likely going to be messy. Instead, I'm thinking of making it a requirement to run the build command from the package's source.

The fix for notifications is thus to make them a package in their own right by moving them to `apps/` and giving them a `package.json` so they can be built through `npx lerna run build` (which is run in the individual source dir): #32643

**Comment by sirreal** [comment]
Calypso.live build fails with a boatload of these: `Missing class properties transform.` for `class { static StaticProp = null; }`

```
Time=2019-04-30T11:13:40.373Z | 
ERROR in ./client/blocks/reader-full-post/index.jsx
Module build failed (from ./node_modules/thread-loader/dist/cjs.js):
Thread Loader (Worker 0)
/calypso/client/blocks/reader-full-post/index.jsx: Missing class properties transform.
  73 | 
  74 | export class FullPostView extends React.Component {
> 75 | 	static propTypes = {
     | 	^
  76 | 		post: PropTypes.object,
  77 | 		onClose: PropTypes.func.isRequired,
  78 | 		referralPost: PropTypes.object,
    at PoolWorker.fromErrorObj (/calypso/node_modules/thread-loader/dist/WorkerPool.js:258:12)
    at File.buildCodeFrameError (/calypso/node_modules/@babel/core/lib/transformation/file/file.js:261:12)
    at NodePath.buildCodeFrameError (/calypso/node_modules/@babel/traverse/lib/path/index.js:157:21)
    at pushBody (/calypso/node_modules/@babel/plugin-transform-classes/lib/transformClass.js:194:20)
    at buildBody (/calypso/node_modules/@babel/plugin-transform-classes/lib/transformClass.js:168:5)
    at classTransformer (/calypso/node_modules/@babel/plugin-transform-classes/lib/transformClass.js:565:5)
    at transformClass (/calypso/node_modules/@babel/plugin-transform-classes/lib/transformClass.js:601:10)
    at PluginPass.ClassExpression (/calypso/node_modules/@babel/plugin-transform-classes/lib/index.js:111:54)
    at newFn (/calypso/node_modules/@babel/traverse/lib/visitors.js:193:21)
    at NodePath._call (/calypso/node_modules/@babel/traverse/lib/path/context.js:53:20)
    at NodePath.call (/calypso/node_modules/@babel/traverse/lib/path/context.js:40:17)
 @ ./client/reader/full-post/controller.js 26:2-28:28 63:2-65:28
 @ ./client/reader/full-post/index.js
 @ ./client/sections.js
 @ ./client/sections-middleware.js
 @ ./client/boot/common.js
 @ ./client/boot/app.js
 @ multi ./client/boot/app

ERROR in ./client/blocks/site/placeholder.jsx
Module build failed (from ./node_modules/thread-loader/dist/cjs.js):
Thread Loader (Worker 0)
/calypso/client/blocks/site/placeholder.jsx: Missing class properties transform.
  10 | 
  11 | export default class extends React.Component {
> 12 | 	static displayName = 'SitePlaceholder';
     | 	^
  13 | 
  14 | 	componentDidMount() {
  15 | 		debug( 'The Site component is mounted.' );
    at PoolWorker.fromErrorObj (/calypso/node_modules/thread-loader/dist/WorkerPool.js:258:12)
    at File.buildCodeFrameError (/calypso/node_modules/@babel/core/lib/transformation/file/file.js:261:12)
    at NodePath.buildCodeFrameError (/calypso/node_modules/@babel/traverse/lib/path/index.js:157:21)
    at pushBody (/calypso/node_modules/@babel/plugin-transform-classes/lib/transformClass.js:194:20)
    at buildBody (/calypso/node_modules/@babel/plugin-transform-classes/lib/transformClass.js:168:5)
    at classTransformer (/calypso/node_modules/@babel/plugin-transform-classes/lib/transformClass.js:565:5)
    at transformClass (/calypso/node_modules/@babel/plugin-transform-classes/lib/transformClass.js:601:10)
    at PluginPass.ClassExpression (/calypso/node_modules/@babel/plugin-transform-classes/lib/index.js:111:54)
    at newFn (/calypso/node_modules/@babel/traverse/lib/visitors.js:193:21)
    at NodePath._call (/calypso/node_modules/@babel/traverse/lib/path/context.js:53:20)
    at NodePath.call (/calypso/node_modules/@babel/traverse/lib/path/context.js:40:17)
 @ ./client/me/concierge/book/skeleton.js 19:0-54 37:29-44
 @ ./client/me/concierge/controller.js
 @ ./client/me/concierge/index.js
 @ ./client/sections.js
 @ ./client/sections-middleware.js
 @ ./client/boot/common.js
 @ ./client/boot/app.js
 @ multi ./client/boot/app
```

**Comment by ockham** [comment]
Thanks Jon! Looks like it's trying to use the wrong Babel config :slightly_frowning_face:

**Comment by ockham** [comment]
The issue can be repro'd locally with `npm run build-client-fallback`.

**Comment by ockham** [comment]
I think that this change is the culprit: https://github.com/Automattic/wp-calypso/pull/32601/files#diff-fc10a6682269c1fefc2abe05d1500b8aR15. `@babel/plugin-proposal-class-properties` is listed as a `plugin` in that preset -- maybe the preset is overridden by others.

However, tying a 'proper' override unfortunately results in the same error :slightly_frowning_face: :

```diff
diff --git a/babel.config.js b/babel.config.js
index 04a41c0d9a..291544594d 100644
--- a/babel.config.js
+++ b/babel.config.js
@@ -11,23 +11,27 @@ const codeSplit = require( './server/config' ).isEnabled( 'code-splitting' );
 // We implicitly use browserslist configuration in package.json for build targets.
 
 const config = {
-	presets: [
-		require( '@automattic/calypso-build/babel/default' ),
-		[
-			'@babel/env',
-			{
-				modules,
-				useBuiltIns: 'entry',
-				corejs: 2,
-				// Exclude transforms that make all code slower, see https://github.com/facebook/create-react-app/pull/5278
-				exclude: [ 'transform-typeof-symbol' ],
-			},
-		],
-	],
+	presets: [ require( '@automattic/calypso-build/babel/default' ) ],
 	plugins: _.compact( [
 		[ '@automattic/transform-wpcalypso-async', { async: isBrowser && codeSplit } ],
 		isBrowser && './inline-imports.js',
 	] ),
+	overrides: [
+		{
+			presets: [
+				[
+					'@babel/env',
+					{
+						modules,
+						useBuiltIns: 'entry',
+						corejs: 2,
+						// Exclude transforms that make all code slower, see https://github.com/facebook/create-react-app/pull/5278
+						exclude: [ 'transform-typeof-symbol' ],
+					},
+				],
+			],
+		},
+	],
 	env: {
 		build_pot: {
 			plugins: [
```

**Comment by ockham** [comment]
Should be finally ready for review :+1:

File: packages/calypso-build/README.md [hunk 1]
```diff
@@ -113,8 +113,8 @@ To extend the default behavior provided by `@automattic/calypso-build`, you can
 ```js
 module.exports = {
 	presets: [
-		require( '@automattic/calypso-build/babel/default' ),
 		require( '@automattic/calypso-build/babel/wordpress-element' ),
+		require( '@automattic/calypso-build/babel/default' ),
```
Comment (sirreal) [inline packages/calypso-build/README.md]: Should these be wrapped in `require`?

Comment (ockham) [reply packages/calypso-build/README.md]: Works with and without `require()`. In some cases, the `require()` provides some safety against accidentally trying to use a preset that doesn't exist (and e.g. ESLint can complain that the package hasn't been added to the project's `package.json`). But yeah, not strictly needed here, and arguably a bit against convention. I'll drop the `require()`s.

Comment (ockham) [reply packages/calypso-build/README.md]: b24083cf54

**Pull request merged**